### PR TITLE
Add initial API docs generation and publishing

### PR DESCRIPTION
**Pull request by sayanshaw24**
Note: all necessary api-docs have been generated and added as of now, if any updates are desired, they can be added in a new PR.

### CODE REVIEW

File: .github/workflows/publish-python-apidocs.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+          sudo apt-get update
+          sudo apt-get install python3-pip
+          sudo apt-get install graphviz
+      - name: Install dependencies
+        run: |
+          python3 -m pip install --upgrade pip
+          python3 -m pip install -r requirements-dev.txt
+          cd docs/python
+          python3 -m pip install -r requirements.txt
+          python3 -m pip install onnxruntime-extensions
```
Comment (wenbingl) [inline .github/workflows/publish-python-apidocs.yml:39]: should it be built and installed from the local repro?

File: docs/python/examples/README.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Gallery of examples
```
Comment (wenbingl) [inline docs/python/examples/README.txt:1]: is this file needed?

File: docs/python/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+coloredlogs
+flatbuffers
+numpy
+packaging
+protobuf
+sympy
+onnx
+sphinx_exec_code
+sphinx_tabs
+furo
```
Comment (wenbingl) [inline docs/python/requirements.txt:23]: duplicated

File: .github/workflows/publish-python-apidocs.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+          python3 -m pip install -r requirements-dev.txt
+          python3 -m pip install -r requirements.txt
+          python3 -m pip install -e .
+          python3 -m pip list
+      - name: Generate Python docs with Sphinx
+        run: |
+          set -e -x
+          python3 -m pip install -r docs/python/requirements.txt
+          rm -rf sphinx_output/docs/
+          python3 -m sphinx -j1 -v -T -b html -d sphinx_output/docs/extensions/_doctrees/html docs/python/ sphinx_output/docs/extensions/html
```
Comment (wenbingl) [inline .github/workflows/publish-python-apidocs.yml:45]: I didn't find where the 'onnxruntime-extensions' package name is specified, did I miss it somewhere?

File: .github/workflows/publish-python-apidocs.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  build:
+    name: Generate Python API docs
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Install tools
+        run: |
+          sudo apt-get update
+          sudo apt-get install python3-pip
+          sudo apt-get install graphviz
```
Comment (wenbingl) [inline .github/workflows/publish-python-apidocs.yml:32]: why is this package needed?

File: .github/workflows/publish-python-apidocs.yml (hunk 4 above)
Comment (sayanshaw24) [reply .github/workflows/publish-python-apidocs.yml:45]: it's in `docs/python/index.rst`

File: docs/python/README.rst [hunk 6]
```diff
@@ -0,0 +1,16 @@
+ONNXRuntime-Extensions
```
Comment (wenbingl) [inline docs/python/README.rst:1]: this information is easily out of date, is this file needed?

File: docs/python/examples/README.txt (hunk 2 above)
Comment (sayanshaw24) [reply docs/python/examples/README.txt:1]: yes, Sphinx requires a README in `examples/` as well to build successfully.

File: docs/python/examples/README.txt
Comment (wenbingl) [inline docs/python/examples/README.txt:1]: can we add examples folder after we really something here?

File: docs/python/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+-f https://download.pytorch.org/whl/torch/
```
[the remark this answers is not in the document]
Comment (sayanshaw24) [reply docs/python/requirements.txt:24]: good catch, i don't think so - will remove if it works successfully without it.

File: docs/python/index.rst [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+.. automodule:: onnxruntime_extensions
+    :members:
+    :undoc-members:
+    :private-members:
+    :special-members:
+    :show-inheritance:
+    :member-order: bysource
+    :inherited-members:
+    :exclude-members:
```
Comment (wenbingl) [inline docs/python/index.rst:17]: Did you try to add 'automodule:: onnxruntime_extensions.tools

File: .github/workflows/publish-python-apidocs.yml (hunk 5 above)
Comment (sayanshaw24) [reply .github/workflows/publish-python-apidocs.yml:32]: we can remove it, it's not needed now since we don't need any graphs for our api-docs.

File: docs/python/index.rst [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+ONNXRuntime-Extensions is a library that extends the capability of the ONNX models
+and inference with ONNX Runtime, via pre/post-processing custom operators.
+Learn more about ONNXRuntime-Extensions `here <https://github.com/microsoft/onnxruntime-extensions>`_.
+
+.. automodule:: onnxruntime_extensions
+    :members:
+    :undoc-members:
+    :private-members:
+    :special-members:
```
Comment (wenbingl) [inline docs/python/index.rst:13]: consider to remove these
    :undoc-members:
    :private-members:
    :special-members:
?

File: docs/python/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,25 @@
+autopep8
+matplotlib
+scikit-learn
+skl2onnx
```
Comment (wenbingl) [inline docs/python/requirements.txt:4]: remove

File: docs/python/examples/README.txt
Comment (sayanshaw24) [reply docs/python/examples/README.txt:1]: unfortunately not, I tried without it, examples folder is required for Sphinx.

File: docs/python/conf.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+import os
+import shutil  # noqa: F401
+import sys
+
+# import recommonmark
+
+# -- Project information -----------------------------------------------------
+
+project = "ONNXRuntime-Extensions Python API"
+copyright = "2018-2023, Microsoft"
```
Comment (skottmckay) [inline docs/python/conf.py:18]: Can we programmatically insert the current year instead of hardcoding 2023?

File: docs/python/README.rst (hunk 6 above)
Comment (sayanshaw24) [reply docs/python/README.rst:1]: nope, removed.

File: docs/python/index.rst (hunk 8 above)
Comment (sayanshaw24) [reply docs/python/index.rst:17]: yes it did not work, so turns out with Sphinx recursive submodule documentation isn't supported properly, so you just need to give it all the modules individually

File: docs/python/conf.py (hunk 11 above)
Comment (sayanshaw24) [reply docs/python/conf.py]: done!

**Review by wenbingl — APPROVED** [review]
